[Temporal] Implement comparison methods for PlainYearMonth

catamorphism opened the pull request:
#### bd18128a9b64b0a779979f28c9de66d52d64242e
<pre>
[Temporal] Implement comparison methods for PlainYearMonth
<a href="https://bugs.webkit.org/show_bug.cgi?id=303799">https://bugs.webkit.org/show_bug.cgi?id=303799</a>

Reviewed by Sosuke Suzuki.

Co-authored-by: SUZUKI Sosuke &lt;aosukeke@gmail.com&gt;

Implement `compare`, `equals`, and `valueOf`.
* JSTests/stress/temporal-plainyearmonth.js:
(shouldBe):
(shouldThrow):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/304196@main">https://commits.webkit.org/304196@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97b1f6d15242924161757adff589ae502d1b6903

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46078 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142342 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86724 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7876 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7110 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103028 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137773 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5560 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120854 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83860 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5393 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3008 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2933 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126864 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39006 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145038 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133333 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6939 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39582 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7002 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5822 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111743 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28357 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5231 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117134 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60871 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6985 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35312 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166235 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6774 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70566 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43445 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7010 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6884 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
```diff
@@ -111,6 +115,25 @@ JSC_DEFINE_HOST_FUNCTION(temporalPlainYearMonthPrototypeFuncToString, (JSGlobalO
     RELEASE_AND_RETURN(scope, JSValue::encode(jsString(vm, yearMonth->toString(globalObject, callFrame->argument(0)))));
 }
 
+// https://tc39.es/proposal-temporal/#sec-temporal.plaindate.prototype.equals
```
```suggestion
// https://tc39.es/proposal-temporal/#sec-temporal.plainyearmonth.prototype.equals
```

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
```diff
@@ -111,6 +115,25 @@ JSC_DEFINE_HOST_FUNCTION(temporalPlainYearMonthPrototypeFuncToString, (JSGlobalO
+    if (yearMonth->plainYearMonth() != other->plainYearMonth())
+        return JSValue::encode(jsBoolean(false));
+
+    return JSValue::encode(jsBoolean(true));
```
Shouldn't we compare calendar here like for `PlainDate#equals` ?

ptomato replied:
On the one hand yes, on the other hand having different calendar values is currently not implemented...

catamorphism replied:
`PlainYearMonth` doesn't have a calendar field yet, because as Philip mentioned, different calendars aren't implemented yet.

sosukesuzuki replied:
Yes, I know that different calendar is not implemented. However, I thought that if it would be needed in the future, it might as well be implemented in this PR for consistency with `PlainDate#equals`.

But I think this is a minor issue and will eventually be implemented and tested, so it is acceptable for me

sosukesuzuki reviewed (approved):
r=me

catamorphism committed:
[Temporal] Implement comparison methods for PlainYearMonth
https://bugs.webkit.org/show_bug.cgi?id=303799

Reviewed by Sosuke Suzuki.

Co-authored-by: SUZUKI Sosuke <aosukeke@gmail.com>

Implement `compare`, `equals`, and `valueOf`.
* JSTests/stress/temporal-plainyearmonth.js:
(shouldBe):
(shouldThrow):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/304196@main
